Add new vector utils (ceil, sign, abs, random_in_area)

kromka-chleba opened the pull request:
Added a few handy vector utils: vector.ceil, vector.sign, vector.abs and vector.random_in_area.
vector.apply got improved to accept optional arguments.

## To do

This PR is Ready for Review

- [x] Code and documentation
- [x] Unit tests

## How to test

Just use the utils I guess? They're so short this shouldn't be a problem.
~~I need to test it myself first too lol.~~

kromka-chleba commented:
Wrote unit tests, now I need to test it in the game.
Otherwise should be ready to review.

kromka-chleba commented:
Okay, everything so far works great.
Do we want `vector.random ` too?

```
function vector.random(...)
    return vector.create(math.random, ...)
end
```

It is useful, but also may be confusing for some. It preserves the default behavior of `math.random`, so if called with no arguments, it would return a vector with xyz: [0, 1), so only positive values by default. The other problematic aspect is that users could use it to create random direction vectors improperly like I did, though #14784 should prevent it.
Otherwise this would be great for generating integer positions randomly, for example.

kromka-chleba commented:
Let me know if you have any other wishes for this PR.
I could also implement `vector.modf` and `vector.fmod` analogous to their `math` counterparts.

grorp commented:
No idea whether `vector.create` is useful, but IMO it needs a better name to distinguish it from `vector.new`. Otherwise people will confuse them.

kromka-chleba commented:
>No idea whether vector.create is useful, but IMO it needs a better name to distinguish it from vector.new. Otherwise people will confuse them.

I use that to implement `vector.random` mentioned above.  It's pretty handy and general enough so people will find uses for it. People should read documentation before using functions, but I can propose different names: `vector.create_with` or `vector.construct_with`.

kromka-chleba commented:
I've been thinking about `vector.create_with` and its usefulness. It was proposed to me by luatic(?) on IRC as a way to generalize my `vector.random` implementation:

```
-- My first proposal
function vector.random(...)
    return vector.new(math.random(...), math.random(...), math.random(...))
end

-- Using the generalized function
function vector.random(...)
    return vector.create_with(math.random, ...)
end
```

So the only use case of `vector.create_with` is creating vectors where each component has to obey some specific rule, yet the components do not depend on each other. So I can't really think of any other uses for it other than implementing `vector.random`. Given the above I could simply remove `vector.create_with`.

appgurueu commented:
I also go by luatic (I don't want to rename on GH (yet?)).

If you can't think of another (at least remotely popular) use case for `vector.create_with`, I'm fine with just replacing it with with `vector.random`.

My thoughts when I proposed this function went vaguely in the direction of non-uniform distributions or pseudorandom, though I suppose we will have a good chunk of use cases covered by `vector.random_direction`, and pseudorandom can be achieved by calling `math.randomseed` and reseeding after you're done (though that is more error-prone than using the engine's pseudorandom since it's easy to forget the reseeding part).

kromka-chleba commented:
Can't think of any real use case honestly. Getting non-symmetrically biased directions would be better done with a function that treats each component separately.
I'll remove it then for now.

kromka-chleba commented:
Okay so I removed `vector.create_with` and added `vector.random_in_area`:

```
function vector.random_in_area(min, max)
	return fast_new(
		math.random(min.x, max.x),
		math.random(min.y, max.y),
		math.random(min.z, max.z)
	)
end
```

I decided not to add `vector.random` as that's a poor and quite useless version of the above, it also allowed implementing code using biased direction.

Otherwise this PR is **ready** now.

kromka-chleba commented:
Rebased with the latest master.

kromka-chleba commented:
Rebased again to fix merge conflict with the vector.random_direction thing.

sfence reviewed (approved):
No problem found.

grorp commented on doc/lua_api.md:
```diff
@@ -3877,6 +3885,12 @@ vectors are written like this: `(x, y, z)`:
+	* Returns a random integer position in area formed by `min` and `max`
+	* `min` and `max` are inclusive.
+	* You can use `vector.sort` if you have two vectors and don't know which are the minimum and the maximum.
+    * Don't use this function for making random direction vectors as it is
+      biased towards the unit cube corners, use `vector.random_direction` instead.
```
Since this returns integers, why would anyone try to use it for generating random directions anyway?

appgurueu replied:
I could imagine modders doing something like `vector.random_in_area(vector.new(-1000, -1000, -1000), vector.new(1000, 1000, 1000)):normalize()` :smile:

(and indeed the unnecessarily increased discretization is another issue with this - though fun fact, lua random usually doesn't have a full mantissa of randomness either)

I agree though that it's probably just as fine to remove this note.

appgurueu reviewed (approved):
I was mildly concerned about the impact of adding a vararg to `vector.apply`, but I did a rudimentary benchmark (`local t = os.clock(); local v = vector.new(1, 2, 3); for i = 1, 1e7 do v = v:apply(math.cos) end; print(v, os.clock() - t)`) and it doesn't seem to be significant.

Side note: `vector.random_in_area(min, max)` is just `vector.combine(min, max, math.random)`
